### DISCUSSION TITLE
Are there plans to support asynchronous APIs?

### DISCUSSION THREAD

**Discussion by J-W-Chan** [abuzuhri/Amazon-SP-API-CSharp#87] [Ideas]
All access to the Amazon API is currently synchronous. Are there plans to support asynchronous APIs? @abuzuhri

**Comment by abuzuhri** [comment]
yes we can do that

**Comment by user6116** [comment]
Yes, I would like this too!

**Comment by abuzuhri** [comment]
Will try to start that soon

**Comment by J-W-Chan** [comment]
You can refer to my branch [https://github.com/J-W-Chan/Amazon-SP-API-CSharp/tree/async-api-request](https://github.com/J-W-Chan/Amazon-SP-API-CSharp/tree/async-api-request)
@abuzuhri

**Comment by abuzuhri** [comment]
@user6116 and @J-W-Chan 

We are fully support asynchronous now

**Comment by user6116** [comment]
So, I have been playing with the Async for some time now.  It's working well"ish".  I think I might be doing something wrong in my code.  Can you explain the proper way that you intended this to be implemented for executing multiple tasks simultaneously?

Let's say I want to run 3 GetPricing's at the same time.  

- Do I create insatiate 3 completely sperate/different Amazon-SP-API objects inside their own classes?    
- Do each of these separate objects use a shared AmazonConnection object, or do they each have their own AmazonConnection object?
- Or would I want only 1 insatiated Amazon-SP-API object and 1 AmazonConnection object and many calls to async methods?

I hope this makes sense.  I have a feeling that my code isn't working right because I'm implementing the async differently than how it was intended to be used.

**Reply by abuzuhri** [reply]
You need to create separate object and each object not share any Amazon connection (there no static )

**Reply by user6116** [reply]
Perfect, thanks!

**Reply by user6116** [reply]
Okay, follow-up question.  I'm getting an error "You exceeded your quota for the requested resource." when I try to spin up too many instances.  I believe your awesome SP-API library has built in support to automatically control the bucket token algorithm.  If I spin up multiple instances, this feature no longer works, is this correct?  Is there any way that you suggest I handle this?

**Reply by abuzuhri** [reply]
For every seller I'd and application you should have one instance only , multi share instance with not work perfect

**Reply by user6116** [reply]
So, I get what you're saying.  But for example the GetMyFeesEstimate api allows 20 calls with a refresh rate of 10 per second.  I couldn't fill that threshold with only 1 instance.  If I want to maximize my processing time, I would need multiple instances running or else I am wasting api calls.  I was thinking of using IsActiveLimitRate to disable the library's built in rate limiter and build my own (cross instance) rate limiter.  Is there any exposed way in the library to access the "x-amzn-RateLimit-Limit" that is read from the header for the specific API calls?

**Reply by abuzuhri** [reply]
Right now no but I can add it in the library

**Reply by user6116** [reply]
Oh wow!  Yeah, that would excellent!  Thanks so much!
